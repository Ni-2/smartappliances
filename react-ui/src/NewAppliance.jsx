import React, { useState } from 'react';
import axios from 'axios';
import cn from 'classnames';

function NewApplianceForm(props) {
  const { serial, changeSerial, setNewApplType, url, setCards } = props;
  const [errorMessage, tryAgain] = useState(false);

  const treatResponse = {
    204: () => {
      tryAgain('Oops! Something is wrong. Please, try again.');
      changeSerial('');
    },
    200: (data) => {
      tryAgain(data);
      changeSerial('');
    },
    201: (data) => {
      setCards(data);
      setNewApplType(data[serial].name);
    },
  };

  const sendQueryForNewAppliance = async (e) => {
    e.preventDefault();
    const res = await axios.post(
      url,
      { serial },
      {
        headers: { 'Content-Type': 'application/json' },
      },
    );
    treatResponse[String(res.status)](res.data);
  };

  return (
    <form className="new-appl-form" onSubmit={sendQueryForNewAppliance}>
      <legend>New appliance</legend>
      <div className="mb-3">
        <label
          htmlFor="serialNumber"
          role={!!errorMessage ? 'invalid-feedback' : 'invitation'}
          className={cn('form-label', { 'invalid-feedback': !!errorMessage }, 'disp-i-b')}
        >
          {!!errorMessage ? errorMessage : 'Enter serial number'}
        </label>
        <input
          className={cn('form-control', { 'is-invalid': !!errorMessage })}
          id="serialNumber"
          aria-describedby="serialHelp"
          value={serial}
          onChange={(e) => changeSerial(e.target.value)}
        />
        <div id="serialHelp" className="form-text">
          The serial number of the appliance is on it's back or bottom side and in the attached
          manual. (( Wash machine: 101, 102, 103; oven: 201, 202, 203; coffee machine: 301, 302, 303
          ))
        </div>
      </div>
      <button type="submit" className="btn btn-primary">
        Submit
      </button>
    </form>
  );
}

function NewApplianceSuccess(props) {
  const { addNewAppliance, toggleToAppliance, serial, newApplType } = props;
  return (
    <div className="alert alert-success" role="alert">
      <h4 className="alert-heading">Well done!</h4>
      <p className="mb-0 flex al-i-base">
        <button
          className="btn btn-link color-unset pad-rev"
          onClick={() => {
            addNewAppliance(false);
            toggleToAppliance(serial);
          }}
        >
          {newApplType}
        </button>
        successfully added to
        <button className="btn btn-link color-unset pad-rev" onClick={() => addNewAppliance(false)}>
          My Appliances.
        </button>
      </p>
    </div>
  );
}

export default function NewAppliance(props) {
  const { addNewAppliance, toggleToAppliance, url, setCards } = props;
  const [serial, changeSerial] = useState('');
  const [newApplType, setNewApplType] = useState(false);

  return (
    <div className="new-appliance-form">
      {newApplType ? (
        <NewApplianceSuccess
          serial={serial}
          newApplType={newApplType}
          addNewAppliance={addNewAppliance}
          toggleToAppliance={toggleToAppliance}
        />
      ) : (
        <NewApplianceForm
          serial={serial}
          changeSerial={changeSerial}
          setNewApplType={setNewApplType}
          url={url}
          setCards={setCards}
        />
      )}
    </div>
  );
}
