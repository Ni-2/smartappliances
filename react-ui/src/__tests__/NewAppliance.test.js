import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import NewAppliance from '../NewAppliance';

const server = setupServer(
  rest.post('/api', (req, res, ctx) => {
    return res(ctx.status(204));
  }),
);

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<NewAppliance />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders form', () => {
  const component = renderer.create(
    <NewAppliance
      addNewAppliance={() => 'function addNewAppliance'}
      toggleToAppliance={() => 'function toggleToAppliance'}
      url="/api"
      setCards={() => 'function setCards'}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('handlers server error', async () => {
  server.use(
    rest.get('/api', (req, res, ctx) => {
      return res(ctx.status(500));
    }),
  );

  render(
    <NewAppliance
      addNewAppliance={() => 'function addNewAppliance'}
      toggleToAppliance={() => 'function toggleToAppliance'}
      url="/api"
      setCards={() => 'function setCards'}
    />,
  );

  fireEvent.click(screen.getByText('Submit'));

  await waitFor(() => screen.getAllByRole('invalid-feedback'));

  expect(screen.getByRole('invalid-feedback')).toHaveTextContent(
    'Oops! Something is wrong. Please, try again.',
  );
});
