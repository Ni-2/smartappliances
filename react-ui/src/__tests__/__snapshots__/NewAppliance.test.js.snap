// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form 1`] = `
<div
  className="new-appliance-form"
>
  <form
    className="new-appl-form"
    onSubmit={[Function]}
  >
    <legend>
      New appliance
    </legend>
    <div
      className="mb-3"
    >
      <label
        className="form-label disp-i-b"
        htmlFor="serialNumber"
        role="invitation"
      >
        Enter serial number
      </label>
      <input
        aria-describedby="serialHelp"
        className="form-control"
        id="serialNumber"
        onChange={[Function]}
        value=""
      />
      <div
        className="form-text"
        id="serialHelp"
      >
        The serial number of the appliance is on it's back or bottom side and in the attached manual. (( Wash machine: 101, 102, 103; oven: 201, 202, 203; coffee machine: 301, 302, 303 ))
      </div>
    </div>
    <button
      className="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
